src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -204,1024 +204,1045 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
                 }
 
-                if (requestParams.IsExtendedLifeTimeEnabled && msalAccessTokenCacheItem.ExtendedExpiresOn >
-                    DateTime.UtcNow + TimeSpan.FromMinutes(DefaultExpirationBufferInMinutes))
+                if (msalAccessTokenCacheItem != null)
                 {
+                    if (msalAccessTokenCacheItem.ExpiresOn >
+                        DateTime.UtcNow + TimeSpan.FromMinutes(DefaultExpirationBufferInMinutes))
+                    {
+                        requestParams.RequestContext.Logger.Info(
+                            "Access token is not expired. Returning the found cache entry. " +
+                            GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+                        return msalAccessTokenCacheItem;
+                    }
+
+                    if (requestParams.IsExtendedLifeTimeEnabled && msalAccessTokenCacheItem.ExtendedExpiresOn >
+                        DateTime.UtcNow + TimeSpan.FromMinutes(DefaultExpirationBufferInMinutes))
+                    {
+                        requestParams.RequestContext.Logger.Info(
+                            "Access token is expired.  IsExtendedLifeTimeEnabled=TRUE and ExtendedExpiresOn is not exceeded.  Returning the found cache entry. " +
+                            GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+
+                        msalAccessTokenCacheItem.IsExtendedLifeTimeToken = true;
+                        return msalAccessTokenCacheItem;
+                    }
+
                     requestParams.RequestContext.Logger.Info(
-                        "Access token is expired.  IsExtendedLifeTimeEnabled=TRUE and ExtendedExpiresOn is not exceeded.  Returning the found cache entry. " +
+                        "Access token has expired or about to expire. " +
                         GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
-
-                    msalAccessTokenCacheItem.IsExtendedLifeTimeToken = true;
-                    return msalAccessTokenCacheItem;
                 }
 
-                requestParams.RequestContext.Logger.Info(
-                    "Access token has expired or about to expire. " +
-                    GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+                return null;
             }
-
-            return null;
         }
-    }
 
-    private string GetAccessTokenExpireLogMessageContent(MsalAccessTokenCacheItem msalAccessTokenCacheItem)
-    {
-        return string.Format(
-            CultureInfo.InvariantCulture,
-            "[Current time ({0}) - Expiration Time ({1}) - Extended Expiration Time ({2})]",
-            DateTime.UtcNow,
-            msalAccessTokenCacheItem.ExpiresOn,
-            msalAccessTokenCacheItem.ExtendedExpiresOn);
-    }
+        private ICollection<MsalAccessTokenCacheItem> FilterToAuthoritySpecifiedByAuthenticationRequest(
```

Comment:
>Authority [](start = 62, length = 9)

nit: TenantID

Reply:
yeah..definitely..thanks

---
In reply to: [240654377](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/714#discussion_r240654377) [](ancestors = 240654377)